### PR TITLE
manager#syncPod should cover reconciliation

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Please see related discussion (around issue #80968): https://github.com/kubernetes/kubernetes/pull/88255#issuecomment-587231603

@yujuhong 's question led me to thinking about whether reconciliation may miss some case.

For the following code in manager#syncBatch :
```
				delete(m.apiStatusVersions, syncedUID)
				updatedStatuses = append(updatedStatuses, podStatusSyncRequest{uid, status})
```
Please note that syncedUID and uid may carry different values.
With this additional log,
```
@@ -517,33 +521,37 @@ func (m *manager) syncBatch() {
                                        continue
                                }
                                syncedUID = mirrorUID
+                               klog.Infof("assigned syncedUID %q uid %q", syncedUID, uid)
                        }
```
Here is one example:
```
I0710 02:47:35.039903    9058 config.go:383] Receiving a new pod "kube-proxy-e2e-9e17d10a0f-a7d53-minion-group-32vz_kube-system(de2df129-bfbd-4c48-90b1-06d49400dad0)"
I0710 02:47:35.040140    9058 kubelet.go:1817] SyncLoop (ADD, "api"): "kube-proxy-e2e-9e17d10a0f-a7d53-minion-group-32vz_kube-system(de2df129-bfbd-4c48-90b1-06d49400dad0)"
...
I0710 02:47:42.007051    9058 status_manager.go:524] assigned syncedUID "de2df129-bfbd-4c48-90b1-06d49400dad0" uid "b31ab15bb4ea18d79cf1cfa40fe53c82"
```

Please note that, in current code, if m.needsUpdate returns false, syncPod() would return early.
However, this may not work with the reconciliation code path (the UID used to check presence in apiStatusVersions in syncPod may be different from the one whose entry is cleared in syncBatch).

This PR allows manager#syncPod to perform reconciliation. Without this change, reconciliation may be skipped (when syncedUID and uid differ).

See https://github.com/kubernetes/kubernetes/pull/88255#issuecomment-599688624 from @byxorna 

Note: another potential fix is to remove uid from apiStatusVersions in manager#syncBatch but this is not as robust as the current fix.

**Which issue(s) this PR fixes**:
Fixes #

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by tedyu** [comment]
The current formation is a little smarter than the first patch @byxorna put in his cluster.

The checkNeedsUpdate flag only needs to be false when m.needsReconcile() says so.

**Comment by derekwaynecarr** [comment]
@smarterclayton i know you are also looking in similar areas.

lets have a broader discussion in sig-node meeting on areas we are investigating on kubelet reliability.  right now a number of the changes are hard to follow in a coherent fashion without a broader story.

**Comment by tedyu** [comment]
@derekwaynecarr 
Should this fix be targeting 1.19 release ?

**Comment by tedyu** [comment]
@kuromt @cwedgwood @mariusgrigoriu
This patch has been running in @byxorna 's cluster for 3 weeks. According to @byxorna this solves the problem for his cluster.

**Comment by xavierW** [comment]
how is it  going ,we came accrose this issue in our production environment

**Comment by haosdent** [comment]
@derekwaynecarr May I know anything block merging this pr. We encounter this problem long long ago and keep backport in our branches. It would be great to see this merged in upstream as well.

**Comment by haosdent** [comment]
verified in our test cluster since March. 
/lgtm

**Comment by smoshiur1237** [comment]
@tedyu  Hi, is there something that is blocking this PR to get merged? We came across this issue in our production environment few months earlier. It would be great to get this PR merged in upstream. is there any possible date  for this PR to get merged?

**Comment by SerialVelocity** [comment]
@derekwaynecarr @smarterclayton 

> lets have a broader discussion in sig-node meeting on areas we are investigating on kubelet reliability. right now a number of the changes are hard to follow in a coherent fashion without a broader story.

Did the discussion ever happen?

**Comment by Quentin-M** [comment]
This bug started breaking various production services last night after some EC2 instances had networking issues.

**Comment by marziman** [comment]
Hello colleagues,

this also started breaking our production from time to time.
Can you please push it, so many folks experiencing issues with this.
@derekwaynecarr @tedyu @dims ?

We would be happy to help testing. Is this targeted to 1.19?

Thansk for all your work!

**Comment by fxposter** [comment]
Are there any progress on this? We see problems in production as well.

**Comment by kirooshu** [comment]
We have been running this patch in production since version 1.17.3 and has resolved the problem for us. We cannot use release Kubernetes until this is merged, unfortunately.

**Comment by shrinandj** [comment]
We're seeing this problem when a readiness gate condition is set on the pod but kubelet is not able to set the pod to be in `Ready` state.

While we're waiting for this fix to be tested, merged and released, are there any suggestions for a workaround? Would it help if the pod had it's readiness gate condition set ALONG WITH some other change (e.g. update to a label/annotation)?

File: pkg/kubelet/status/status_manager.go
Comment (smarterclayton) [inline pkg/kubelet/status/status_manager.go:545]: `m.needsUpdate` should be hoisted up and out of syncPod rather than passing in a boolean.

**Review by smarterclayton — CHANGES_REQUESTED** [review]

File: pkg/kubelet/status/status_manager.go
Comment (tedyu) [reply pkg/kubelet/status/status_manager.go:545]: needsUpdate is not a parameter of the func.

If you meant checkNeedsUpdate, I think it should be a parameter instead of a field.

I am open to better naming of the parameter.

Comment (mmerkes) [reply pkg/kubelet/status/status_manager.go:545]: Would it be less confusing if you flipped the boolean around? I.e. `forceSync`, `alwaysSync`?

I could be wrong, but I think @smarterclayton is suggesting that rather than hiding the fact that we might skip the sync, pull the `m.needsUpdate(uid, status)` out of `syncPod` and do that check before calling `syncPod`. I agree that it's more convenient to not add `if (m.needsUpdate(uid, status)) { m.syncPod(update.podUID, update.status) }` everywhere we call `syncPod`, but it doesn't make `syncPod` cleaner and get rid of a variable that's always going to be a little confusing.

Comment (tedyu) [reply pkg/kubelet/status/status_manager.go:545]: Extracting the check to ahead of syncPod() calls seems to be bigger change.
Several places need to be updated, e.g.
```
pkg/kubelet/pod_workers.go:
                     err = p.syncPodFn(syncPodOptions{
```
If possible, I want to leave such refactoring to subsequent PRs.

I am open to renaming checkNeedsUpdate, once more people chime in and we get consensus on the final variable name.

thanks

Comment (mmerkes) [reply pkg/kubelet/status/status_manager.go:545]: IMO, that's refactoring that nobody's ever likely to do, so we probably just need consensus here.

As for the several places that need updating, I don't believe it would impact anywhere else. `syncPod` is private, so it's not called outside of this file. The place that you reference in `pkg/kubelet/pod_workers.go` seems to trace back to [this method](https://github.com/kubernetes/kubernetes/blob/dd466bccde8176bd390fcf712c0752ae94444742/pkg/kubelet/kubelet.go#L1398), which calls StatusManager methods that take care of that.

Comment (tedyu) [reply pkg/kubelet/status/status_manager.go:545]: It seems to me that leaving the needsUpdate() call inside syncPod() would result in slightly shorter code.
In the future, if one more call to syncPod is added somewhere, the developer wouldn't forget that needsUpdate() may be involved.

Comment (mmerkes) [reply pkg/kubelet/status/status_manager.go:545]: Shorter code is something to strive for, but doesn't take precedence over readability. `needsUpdate()` is an optimization, if I understand it correctly, so the risk that someone doesn't know to check `needsUpdate()` seems low and they'll have the other uses as examples.

Comment (tedyu) [reply pkg/kubelet/status/status_manager.go:545]: I am flexible w.r.t. how to refactor the check.

I would prefer confirmation from maintainers that the fix itself is correct.

Comment (smarterclayton) [reply pkg/kubelet/status/status_manager.go:545]: In general in the refactorings I was doing syncPod is already wrong, and passing a boolean to do an if check to a function is usually wrong.  Do the check above the function call, because syncPod is already supposed to be called in limited spots.

Comment (mmerkes) [reply pkg/kubelet/status/status_manager.go:545]: @tedyu You have time to update this PR?

**Comment by mitchelldavis44** [comment]
Bump, any status on this?  This is obviously a major issue for a lot of people.

File: pkg/kubelet/status/status_manager.go
Comment (mmerkes) [reply pkg/kubelet/status/status_manager.go:545]: @tedyu Let me know if you don't have the bandwidth to make the requested changes and get this over the line. If not, I can try to do it.

**Comment by gtorre** [comment]
I agree with @mitchelldavis44

**Comment by tedyu** [comment]
I looked at review comments so far. I think the current form of the patch is good to go: it has been verified by various deployments (back in Apr 2020).

**Comment by tedyu** [comment]
@ehashman 
Can you tell me what you plan to do with this PR ?

**Comment by ehashman** [comment]
@tedyu I assigned the PR so I can track it, as I wanted to review it in detail. I'm looking into this to see if this PR might fix some race issues we've been seeing in CI -- are you planning on continuing to work on this PR? If not, I might open another one based on top of these changes.

**Comment by tedyu** [comment]
Since this is an important fix, I would prefer the commit would bear my name.

I have some bandwidth for upcoming review comments.

**Comment by ehashman** [comment]
> Since this is an important fix, I would prefer the commit would bear my name.

If I pick up your branch I won't modify the commit, I will just add on top of it.

**Comment by jxl4650152** [comment]
Hey, 
is there a workaround for this situation? Our production environment encountered the same problem.

**Comment by tedyu** [comment]
@jxl4650152
I am not aware of a workaround.

**Review by ehashman — CHANGES_REQUESTED** [review]
Please don't fight the bot on stale PRs. 

There are no accompanying test changes for this PR so we can't be sure it actually addresses the issue. We also recently refactored the PLEG so this change may no longer be necessary. Without demonstrated impact I am inclined to close this PR.

**Comment by tedyu** [comment]
I would suggest giving this PR some time to see if the problem still occurs in latest release.

**Comment by tedyu** [comment]
Which old releases did the PLEG refactoring go to ?

What about the releases without the refactoring ? Shouldn't those releases be covered (by this fix)?

**Comment by tedyu** [comment]
I took a look at recent checkins to pkg/kubelet/status/status_manager.go

As far as I can tell, the logic touched by the PR remains unchanged.

@ehashman 
Can you point me to the commits which you referred to earlier ?

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Comment by CormickKneey** [comment]
> Please don't fight the bot on stale PRs.
> 
> There are no accompanying test changes for this PR so we can't be sure it actually addresses the issue. We also recently refactored the PLEG so this change may no longer be necessary. Without demonstrated impact I am inclined to close this PR.

Any proposal doc or reference link can be followed?

**Pull request closed without merging**
